awx/main/tasks/jobs.py

```diff
@@ -1324,62 +1334,23 @@ def clear_project_cache(cache_dir, keep_value):
                     except OSError:
                         logger.warning(f"Could not remove cache directory {old_path}")
 
+    def make_local_copy_with_lock(self, project, job_private_data_dir):
+        self.acquire_lock(project)
+        try:
+            self.make_local_copy(project, job_private_data_dir)
+        finally:
+            self.release_lock(project)
```

Comment:
I'm still not happy with this one part. It's superficial and doesn't really work.

We need to obtain the lock before we check the revision, and we need to keep holding it until after the folder copy is done.

```python
git_repo = git.Repo(project_path)
if job_revision == git_repo.head.commit.hexsha:
  ...
```

Otherwise there's not really any point in this. I didn't really want to make a context manager out of this, but maybe that's what will happen.